From b2b282ca7b46a8199a7a6029bc66a1ced4cd63e8 Mon Sep 17 00:00:00 2001
From: nxa13443 <chaofan.huang@nxp.com>
Date: Thu, 29 Mar 2018 18:58:04 +0800
Subject: [PATCH 3551/5242] MLK-17945[IMX8QXP B0] Deal with the case that
 firmware download fails

commit  bf92f323ff93ec1ac350018edc3c38073fc5605b from
https://source.codeaurora.org/external/imx/linux-imx.git

Deal with the case that vpu firmware download fails, release the context
data

Signed-off-by: nxa13443 <chaofan.huang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mxc/vpu-decoder-b0/vpu_b0.c         |   11 ++++++++++-
 drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c |   11 ++++++++++-
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/drivers/mxc/vpu-decoder-b0/vpu_b0.c b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
index 0442065..ec5d796 100755
--- a/drivers/mxc/vpu-decoder-b0/vpu_b0.c
+++ b/drivers/mxc/vpu-decoder-b0/vpu_b0.c
@@ -2105,7 +2105,7 @@ static int v4l2_open(struct file *filp)
 		if (ret) {
 			vpu_dbg(LVL_ERR, "error: vpu_firmware_download fail\n");
 			mutex_unlock(&dev->dev_mutex);
-			return ret;
+			goto err_firmware_load;
 		}
 		dev->fw_is_ready = true;
 	}
@@ -2149,6 +2149,15 @@ static int v4l2_open(struct file *filp)
 				__func__, ctx->udata_buffer_size, ctx->udata_buffer_virt, (void *)ctx->udata_buffer_phy);
 
 	return 0;
+
+err_firmware_load:
+	release_queue_data(ctx);
+	ctrls_delete_decoder(ctx);
+	v4l2_fh_del(&ctx->fh);
+	v4l2_fh_exit(&ctx->fh);
+	clear_bit(ctx->str_index, &dev->instance_mask);
+	kfree(ctx);
+	return ret;
 }
 
 static int v4l2_release(struct file *filp)
diff --git a/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c b/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
index 8d8609f..dd6ba17 100644
--- a/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
+++ b/drivers/mxc/vpu-encoder-b0/vpu_encoder_b0.c
@@ -1542,7 +1542,7 @@ static int v4l2_open(struct file *filp)
 		if (ret) {
 			vpu_dbg(LVL_ERR, "error: vpu_firmware_download fail\n");
 			mutex_unlock(&dev->dev_mutex);
-			return ret;
+			goto err_firmware_load;
 		}
 		dev->fw_is_ready = true;
 	}
@@ -1579,6 +1579,15 @@ static int v4l2_open(struct file *filp)
 	ctx->actFrame.size = 0;
 
 	return 0;
+
+err_firmware_load:
+	release_queue_data(ctx);
+	ctrls_delete_encoder(ctx);
+	v4l2_fh_del(&ctx->fh);
+	v4l2_fh_exit(&ctx->fh);
+	clear_bit(ctx->str_index, &dev->instance_mask);
+	kfree(ctx);
+	return ret;
 }
 
 static int v4l2_release(struct file *filp)
-- 
1.7.9.5

